src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -885,98 +987,114 @@ internal static void PerformWaitOrTimerCallback(_ThreadPoolWaitOrTimerCallback h
+             )
         {
             if (millisecondsTimeOutInterval < -1)
                 throw new ArgumentOutOfRangeException(nameof(millisecondsTimeOutInterval), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
             return RegisterWaitForSingleObject(waitObject, callBack, state, (uint)millisecondsTimeOutInterval, executeOnlyOnce, false);
         }
 
-        public static RegisteredWaitHandle RegisterWaitForSingleObject(
+        public static RegisteredWaitHandle RegisterWaitForSingleObject(  // throws RegisterWaitException
             WaitHandle waitObject,
             WaitOrTimerCallback callBack,
             object state,
             long millisecondsTimeOutInterval,
-            bool executeOnlyOnce)
+            bool executeOnlyOnce    // NOTE: we do not allow other options that allow the callback to be queued as an APC
+        )
         {
-            if (millisecondsTimeOutInterval < -1 || millisecondsTimeOutInterval > int.MaxValue)
+            if (millisecondsTimeOutInterval < -1)
                 throw new ArgumentOutOfRangeException(nameof(millisecondsTimeOutInterval), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
             return RegisterWaitForSingleObject(waitObject, callBack, state, (uint)millisecondsTimeOutInterval, executeOnlyOnce, true);
         }
 
-        public static RegisteredWaitHandle UnsafeRegisterWaitForSingleObject(
+        public static RegisteredWaitHandle UnsafeRegisterWaitForSingleObject(  // throws RegisterWaitException
             WaitHandle waitObject,
             WaitOrTimerCallback callBack,
             object state,
             long millisecondsTimeOutInterval,
-            bool executeOnlyOnce)
+            bool executeOnlyOnce    // NOTE: we do not allow other options that allow the callback to be queued as an APC
+        )
         {
-            if (millisecondsTimeOutInterval < -1 || millisecondsTimeOutInterval > int.MaxValue)
+            if (millisecondsTimeOutInterval < -1)
                 throw new ArgumentOutOfRangeException(nameof(millisecondsTimeOutInterval), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
             return RegisterWaitForSingleObject(waitObject, callBack, state, (uint)millisecondsTimeOutInterval, executeOnlyOnce, false);
         }
 
         public static RegisteredWaitHandle RegisterWaitForSingleObject(
-            WaitHandle waitObject,
-            WaitOrTimerCallback callBack,
-            object state,
-            TimeSpan timeout,
-            bool executeOnlyOnce)
+                          WaitHandle waitObject,
+                          WaitOrTimerCallback callBack,
+                          object state,
+                          TimeSpan timeout,
+                          bool executeOnlyOnce
+                          )
         {
-            int tm = WaitHandle.ToTimeoutMilliseconds(timeout);
+            long tm = (long)timeout.TotalMilliseconds;
```

Comment:
This looks like a bug fix. It would be useful to submit it in a separate PR + matching tests in CoreFX.

Reply:
Not really a bugfix. I just took the CoreCLR version since it doesn't have `WaitHandle.ToTimeoutMilliseconds` internal method. It fixes the error message for `tm > (long)int.MaxValue` case, but the exception type and conditions are the same.